src/executor/nested_loop_join.rs

```diff
@@ -21,170 +21,94 @@ pub struct NestedLoopJoinExecutor {
-                                    }
-                                    None => {}
-                                }
-
-                                for (idx, builder) in chunk_builders.iter_mut().enumerate() {
-                                    builder.push(&left_row[idx]);
-                                }
-                            }
-                            DataValue::Bool(false) => {}
-                            _ => {
-                                panic!("unsupported value from join condition")
-                            }
-                        }
-                    }
+                for _ in right_rows() {
+                    matched |= matches!(filter.get(i), Some(true));
+                    i += 1;
                 }
-                match &mut left_bitmaps {
-                    Some(left_bitmaps) => {
-                        if matched {
-                            left_bitmaps[left_chunk_idx].set(left_row_idx, true);
-                        }
-                    }
-                    None => {}
+                if matched {
+                    continue;
+                }
+                // append row: (left, NULL)
+                let values =
+                    (left_row.values()).chain(self.right_types.iter().map(|_| DataValue::Null));
+                for (builder, v) in builders.iter_mut().zip(values) {
+                    builder.push(&v);
                 }
             }
+            yield builders.into_iter().collect();
         }
-        match &left_bitmaps {
-            Some(left_bitmaps) => {
-                for left_chunk_idx in 0..left_chunks.len() {
-                    for left_row_idx in 0..left_chunks[left_chunk_idx].cardinality() {
-                        if !left_bitmaps[left_chunk_idx][left_row_idx] {
-                            let mut left_row =
-                                left_chunks[left_chunk_idx].get_row_by_idx(left_row_idx);
-                            for _ in 0..right_row_len {
-                                left_row.push(DataValue::Null);
-                            }
 
-                            for (idx, builder) in chunk_builders.iter_mut().enumerate() {
-                                builder.push(&left_row[idx]);
-                            }
-                        }
+        // append rows for right outer join
+        if matches!(
+            self.join_op,
+            BoundJoinOperator::RightOuter | BoundJoinOperator::FullOuter
+        ) {
+            let mut builders = create_builders();
+            // scan the filter to find whether a right row matches any left rows
+            let mut matched = bitvec![0; right_rows().count()];
+            let mut i = 0;
+            for _ in left_rows() {
+                for (j, _) in right_rows().enumerate() {
+                    if matches!(filter.get(i), Some(true)) {
+                        matched.set(j, true);
                     }
+                    i += 1;
```

Comment:
Why maintain I manually?

Reply:
I need something like `.enumerate()` in the nested for-loop. This way is the simplest I can think (

Reply:
got it

Reply:
How about https://docs.rs/itertools/0.4.8/itertools/trait.Itertools.html#method.cartesian_product?

Reply:
catesian_product().enumerate(), then we may got i?

Not necessary, any one is ok

Reply:
I tried:
```rust
for (i, (_, (j, _))) in left_rows()
    .cartesian_product(right_rows().enumerate())
    .enumerate()
{
    if matches!(filter.get(i), Some(true)) {
        matched.set(j, true);
    }
}
```

However, `.cartesian_product(other)` requires `other` to be cloneable, unfortunately that is impossible in our case.

Give up 🤪

Reply:
I'm stupid. It works now:
```rust
for (i, (_, j)) in left_rows()
    .cartesian_product(0..right_rows().count())
    .enumerate()
{
    if matches!(filter.get(i), Some(true)) {
        matched.set(j, true);
    }
}
```

Reply:
I found another way which can avoid bitmap: a46215c